agentcfg/cache.go

```diff
@@ -31,53 +33,38 @@ const (
 
 type cache struct {
 	logger  *logp.Logger
-	exp     time.Duration
 	gocache *gocache.Cache
 }
 
 func newCache(logger *logp.Logger, exp time.Duration) *cache {
-	if logger == nil {
-		logger = logp.NewLogger("agentcfg")
-	}
 	logger.Infof("Cache creation with default expiration %v.", exp)
 	return &cache{
 		logger:  logger,
-		exp:     exp,
 		gocache: gocache.New(exp, cleanupInterval)}
 }
 
-func (c *cache) fetchAndAdd(q Query, fn func(Query) (*Doc, error)) (doc *Doc, err error) {
-	id := q.ID()
-
+func (c *cache) fetch(q Query, fetch func() (Result, error)) (Result, error) {
 	// return from cache if possible
-	doc, found := c.fetch(id)
-	if found {
-		return
+	value, found := c.gocache.Get(q.Etag)
+	if found && value != nil {
+		return value.(Result), nil
 	}
-
-	// call fn to retrieve resource from external source
-	doc, err = fn(q)
+	// retrieve resource from external source
+	result, err := fetch()
 	if err != nil {
-		return
+		return result, err
 	}
-
-	// add resource to cache
-	c.add(id, doc)
-	return
-}
-
-func (c *cache) add(id string, doc *Doc) {
-	c.gocache.SetDefault(id, doc)
-	if !c.logger.IsDebug() {
-		return
+	if !authorized(q.IsRum, result.Source.Agent) {
+		return ZeroResult(), nil
```

Comment:
With this implementation non authorized RUM results are not cached. IMO this should be changed.

Reply:
I don't think so. If then the same request comes trough the right endpoint, the result would be a wrong empty result.

Reply:
Kibana is responding the same way to this query, no matter who requests it, right? So I don't see why caching the `result` here would lead to wrong results.
Ofc I did not mean to cache the `ZeroResult` but the `result`.

Reply:
in that case i don't know why i would want to cache something that i am not even returning, but im fine either way

Reply:
My reasoning is the following:
What is the purpose of caching here? 
To avoid unnecessary calls to Kibana based on legitimate requests from agents. But as soon as those requests can be triggered from requests to unsecured RUM endpoint, also to prevent overloading Kibana on any kind of attacks.
The `authorized` check would only return `false` if the agents had a bug or for malicious requests. In the case of malicious requests it makes sense to also avoid sending the same requests over and over to Kibana. 

However, if you add it to the cache, you would probably also need to size limit the cache, as otherwise you open up another attack vector.

I don't see why the decision in which cases something should be added to the cache is in direct relation on whether or not you are returning this information.

Reply:
Not sure you find this useful, but I gave this some thought when introducing the caching https://github.com/elastic/apm-server/issues/2220#issuecomment-504935447.

Reply:
Thanks. 

I am changing the caching key to be service name + env, instead of Etag.
When there are many RUM agents, they will not send an `ifnonematch` arg until they get one from a previous call, which means that all of them would have to hit Kibana at least one.
Caching the service name solves this problem. 

On the other side, I don't think is necessary to have a cache limit.

Reply:
cd40ddbdc4af56a6cf941c9a377cccb5c3f05ea2